Replace llvm.memcpy et al's i1 isVolatile with i8 VolFlags

urnathan opened the pull request:
Modify memcpy, memcpy.inline & memmove intrinsics to change the final i1 isVolatile parameter becomes an i8 VolFlags parameter. The former is true if either (or both) dst and src point to volatile storage. The new VolFlags separates dst and src volatilities to separate bits in the argument value. This allows any expansion code to only mark the volatile accesses as volatile. The motivating use is copying a volatile structure to/from non-volatile storage.

The patch series:

    * Changes the intrinsics, documentation and AutoUpdater. The APIs are not updated, and the existing APIs convert false/true to 0/3 as required.
    * Update MLIR
    * Update Flang
    * Update Polly (tests only)
    * At this point the compiler builds, but tests fail
    * Update LLVM & Clang non-UTC test cases
    * Update LLVM & Clang UTC test cases
    * At this point the conversion is done, but nothing makes use or can generate the new values.
    * Update IRBuilder APIs to specify separate Dst & Src Volatilities. The API change is backwards compatible and a deprecation warning is added in a later patch (it's quite noisy until the remaing conversions are done).
    * Update Clang code gen to specify separate Dst & Src Volatilities.
    * At this point IRBuilder and Clang are generating more precise volatility information, but no target lowering makes use of it.
    * Update TargetLowering APIs. New predicates added to MemTransferInst, and as with the above API change, this is source compatible, but will eventually educes deprecation warnings.
    * Update llvm transforms and target lowering APIs. New EmitTargetCodeForMem{cpy,move} added to SelectionDagTargetInfo, source compatible by forwarding to the old API.
    * Target-specific changes for the new APIs.
    * Motivating testcase.
    * Convert remaining old-API uses, but mark (some) locations as candidates for further optimization opportunities.
    * Deprecate the compatibility APIs, this alerts out-of-tree maintainers.
    * Conversion of AArch64's bespoke memcpy/memmove expansion to use the new EmitTargetCodeForMem{cpy,move} API.

The MLIR and Flang conversions took about 4 hours each -- not having previously looked at those code bases. Thus I believe any out-of-tree conversions will be straight forwards too. Just pay attention to the deprecation warnings and pass the correct tuple.

90 % of this diff are mechanical testcase changes (either via sed or UTC). Most of the rest is fairly mechanical, located via deprecation warnings.

urnathan committed:
[llvm] Modify memcpy intrinsics to take dst/src volatile flags

Modify memcpy, memcpy.inline & memmove intrinsics to change the final
i1 isVolatile parameter to be an i8 VolFlags parameter. The former is
true if either (or both) dst and src point to volatile storage. The
new VolFlags separates dst and src volatilities to separate bits in
the argument value. This allows any expansion code to only mark the
volatile accesses as volatile. The motivating use is copying a
volatile structure to/from non-volatile storage. We get better code
generation when expanding such one-sided memcpys, as the non-volatile
side can (often) be held in registers.

Other than changing the intrinsics, the APIs are not adjusted in this patch.

urnathan committed:
[llvm] Adjust IR Builder for memcpy/memmove volatile change

Introduce llvm::MemTransferVolatility. This provides separate Dst &
Src volatility flags, along with compatibility functions to convert
from and to a single bool. Those will be deprecated in a later
patch, so uses generate warnings.

The memcpy et al builders now take a MemTransferVolality, and encode
it into the VolFlags argument value.

urnathan committed:
[llvm] Update manual memcpy/move intrinsic testcases

Update the non-UTC llvm testcases for memcpy/memmove intrinsic
isVolatile->VolFlags change. Only the check lines are altered --
existing llvmir using the old boolean is left and altered during read
in.

urnathan committed:
[llvm] Update UTC tests for memcpy/memmove intrinsic change

Regenerate llvm UTC tests for the isVolatile->VolFlags change.

urnathan committed:
[llvm] Remaining memcpy/memmove volatile argument cleanup

Silence the remaining places where distinct dst/src volatility flags
are now required. I suspect several of these are opportunities for
further optimization work, thus marking with FIXMEs.

urnathan committed:
Volatile structure memcpy test case

The motivating testcase for memcpy/memmove VolFlags change. Copying a
volatile strucure to/from non-volatile storage should not pessimize
the latter. With the new flags, we can hoist the non-volatile object
into registers and process it directly there.

urnathan committed:
[mlir] Updates for memcpy/memmove intrinsic change

This update MLIR for the intrinsics isVolatile->VolFlags change. The
MLIR parameter name remains isVolatile.

urnathan committed:
[flang] Updates for memcpy/memmove intrinsics change

Fortran has no concept of 'volatile', but this updates Flang's IR to
represent memcpy/memmove's new VolFlags argument. We only pass i8:0
(replacing the previous i1:false).

urnathan committed:
[clang] Update UTC tests for memcpy/memmove intrinsic change

Regenerate clang UTC tests for the isVolatile->VolFlags change. Most
of these were first generated via sed, and then regenerated via
UTC.

urnathan committed:
[llvm] MemTransferInst change for memcpy/memmove volatility

Add separate dst & src volatility accessors to MemTransferInst. Add
MemTransferVolatility to selection dag's getMemcpy,
getMemmove. Existing uses are source compatible, but that API will be
deprecated in a later patch to catch any remaining uses.

urnathan committed:
[aarch64] Update memcpy et al emission for separate volatilities

Update AArch64SelectionDAGInfo::EmitTargetCodeForMem{cpy,move} to override the
separate VolFlags API.

urnathan committed:
[clang] Update CodeGen for memcpy/memmove volatile changes

Update Clang's generation of memcpy et al intrinsics to specify
separate Dst & Src volatilities. Completes Clang's conversion to the
new API.

urnathan committed:
[llvm] CodeGen & Transform updates for memcpy/memmove volatility

Replace single isVolatile with VolFlags. Add a new
SelectionDAGTargetInfo::EmitCodeForMem{Cpy,Move} virtual function with
the new API. The default implementation of which forwards to the old
API.

urnathan committed:
[llvm] Target-specific changes for memcpy/memove volatility

Update all the Target code generation to use the new APIs. Does not
convert the EmitTargetCodeForMem{cpy,move} implementations.

urnathan committed:
[clang] Update manual memcpy/move intrinsic testcases

Update the non-UTC clang testcases for memcpy/memmove intrinsic
isVolatile->VolFlags change. Only the check lines are altered --
existing llvmir using the old boolean is left and altered during read
in.

urnathan committed:
[llvm] Deprecate Volatility compatibility API

Deprecate the compatibilty API of the new dst/src volatility
flags. There are no remaining uses, but of course out-of-tree ports
are a thing.

urnathan committed:
[polly] Test fixes for memcpy/memmove intrinsics change

joker-eph commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -214,14 +214,14 @@ class LLVM_MemcpyIntrOpBase<string name> :
     /*requiresAccessGroup=*/1, /*requiresAliasAnalysis=*/1> {
   dag args = (ins Arg<LLVM_AnyPointer,"",[MemWrite]>:$dst,
                   Arg<LLVM_AnyPointer,"",[MemRead]>:$src,
-                  AnySignlessInteger:$len, I1Attr:$isVolatile);
```
Seems to me that should handle this as we do with the FastMathFlags: that is a enum attribute modeling the bitfields.

urnathan replied:
thanks, looking ...

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -24790,8 +24790,8 @@ static SDValue LowerVACOPY(SDValue Op, const X86Subtarget &Subtarget,
       DAG.getIntPtrConstant(Subtarget.isTarget64BitLP64() ? 24 : 16, DL),
-      Align(Subtarget.isTarget64BitLP64() ? 8 : 4), /*isVolatile*/ false, false,
-      false, MachinePointerInfo(DstSV), MachinePointerInfo(SrcSV));
+      Align(Subtarget.isTarget64BitLP64() ? 8 : 4), /*Vol=*/{false, false},
```
For understandability, should we use a MemTransferVolatility builder helper instead of embedding implicit values like this?

urnathan replied:
good idea, something like `MemTransferVolatility().setDst(true/false).setSrc(true/false)` ? Do you have preference for names?

RKSimon replied:
No preference for names - do we need to handle runtime bool values or could a few basic hardcoded combos be enough MemTransferVolatility::neverVolatile() / MemTransferVolatility::readVolatile() ?

urnathan replied:
Just pushed a patch that uses `MemTransferVolatility` default ctor for the non-volatile case, and allows `.Dst(bool)`, `.Src(bool)` and/or `.All(bool)` to explicitly specify.  One does need the ability to provide runtime values there.

urnathan committed:
Add Builder API to MemTransferVolatility

Rather than permit {bool, bool}, use
MemTransferVolatility().Dst(bool).Src(bool).All(bool). One can only
set bits this way, not clear them.

nikic commented:
Related discourse thread: https://discourse.llvm.org/t/copy-to-from-volatile-structure/72278
